index.rst

```diff
@@ -1,62 +1,427 @@
+Its implementation has been blocked by a lack of butler query-system functionality (essentially :ref:`feature-data-id-set-upload`) that we have been unable to prioritize.
+
+This algorithm is relevant for Campaign Management/Definition because it allows filters on data IDs - whether provided by data ID sets or boolean contraint expressions - to be specific to certain tasks or dataset types, for example allowing one data ID set to be used for one kind of coadd, and another data ID set to be used for a different type of coadd.
+Without this, the only way to have different tasks operate on different sets of input data IDs is via task code in either `PipelineTaskConnections.adjustQuantum` or `PipelineTask.runQuantum`.
+
+.. _other-drivers:
+
+Other Drivers for Middleware Features
+=====================================
+
+Many of the features described here are important for other DM needs, and when considering the total development cost of a new Campaign Management/Definition, it makes sense to consider these "discounted" (or at least easier to prioritize) at some level.
+
+- `DM-31725 <https://jira.lsstcorp.org/browse/DM-31725>`__ has come up repeatedly here as a blocker for other features.
+  It also blocks vectorized calibration-dataset lookup (the absence of which is frequently the bottleneck in `QuantumGraph` generation) and support for `PipelineTasks` whose dimensions include HEALPix or HTM (such as a task to produce HiPS maps).
+
+- Without :ref:`feature-per-task-quantum-graph-generation` / `DM-21904 <https://jira.lsstcorp.org/browse/DM-21904>`__, some tasks in the DRP pipeline cannot safely be run as part of the same submission, forcing the pipeline to be split up into more steps than we would like.
+
+- :ref:`feature-dynamic-dimensions` / `DM-33751 <https://jira.lsstcorp.org/browse/DM-33751>`__ is one of two possible solutions to the problem of how the image cutout service should identify its output datasets (the other is allowing some dataset types to have non-unique data IDs within a run, which is less generally useful).
+  We expect this to be useful more broadly in Science Platform services and user-defined processing, which is likely to want data ID keys other than those needed for our own processes.
+
+- :ref:`feature-quantum-provenance` plays a key role in satisfying fundamental middleware requirements.
+
+.. _summary-and-recommendations:
+
+Summary and Recommendations
+===========================
+
+Recommendations
+---------------
+
+Based on current middleware capabilities, the perceived difficulty of extending those capabilities in the ways described above, and our guesses at how difficult it would be to stand up and maintain non-middleware (or less-middleware-based) solutions to various problems, we make the following recommendations for Campaign Management/Definition:
+
+#. We should mandate a one-to-one relationship between ``RUN`` collections and workflows.
```

Comment:
That's a good point, and I don't have a solid answer right now.  Certainly the right thing to do to get any per-RUN datasets is to use the RUN to query for them, and use the CHAINED collection only to find the RUNs, which is possible (or in the future, we could use quantum provenance to get them). But I agree it would be nice to make it hard to accidentally get only some arbitrary, semi-latest one by querying the CHAINED collection directly.  I can think of a few possible solutions that wouldn't be hard to implement, but would also involve some low-level special-casing, and I'd prefer to come up with a more general mechanism if I can.